documentation/API-design-guidelines.md

```diff
@@ -1215,7 +1216,7 @@ The `subscriptionDetail` must have at least an eventType attribute:
 
 | name | type | attribute description | cardinality |
 | ----- |	-----  |	 -----  |  -----  | 
-| eventType | string | Type of event subscribed. This attribute must be present in the `POST` request. It is open to API working group to allow providing a list of event type based on specific UC. `eventType` must follow UPPER_SNAKE_CASE format. | mandatory  |
+| subject | string | Type of event subscribed. This attribute must be present in the `POST` request. It is open to API working group to allow providing a list of event subject based on specific UC. `subject` must follow CamelCase format. | mandatory  |
```

Comment:
@PedroDiez 

Thanks for the review. As a lot of your comments are about `type` and `subject` usage I tried here to explain our thinking.

Using `type` attribute was our initial approach (with @patrice-conil) but crafting the device status API we got a problem with this approach and proposed to switch to `subject` (to make the thing simpler for implementation & consumers).

Let me explain why:

- Accordingly to cloudEvents [specification ](https://github.com/cloudevents/spec/blob/main/cloudevents/spec.md) the `type` contains a value describing the type of event related to the originating occurrence and should be prefixed with a reverse-DNS name. So the value should be something like this `org.camara.api.device-status.RoamingStatusEvent`. 
- For a lot of API we will have several types of event. In the swagger of the API we need to provide for each event type a data structure. This data structure extend event structure with specific attribute related to the event type. In order to do that we use extension pattern with discriminator 
-  With the `type` value it seems to us too complex to use this `type` field as a `discriminator` - that's why we proposed to use `subject` to store the event type.  This is valid approach as cloudEvents specification define subject as "_the subject of the event in the context of the event producer_" 
- Last but not least as this _subject_ is use a discriminator it made sense for us to use same format (CamelCase) as the one used in the swagger for the classes.  

We crafted the device status API with this new pattern and the CloudEvent class definition will be:

```yaml
    CloudEvent:
      description: The notification callback
      required:
        - id
        - source
        - specversion
        - type
      properties:
        id:
          type: string
          description: identifier of this event, that must be unique in the source context.
        source:
          $ref: '#/components/schemas/Source'
        type:
          type: string
          description: 'type of event as defined in each CAMARA API'
          example: org.camara.api.device-status.RoamingStatusEvent
        specversion:
          type: string
          description: Version of the specification to which this event conforms (must be 1.0 if it conforms to cloudevents 1.0.2 version)
          example: 1.0
        datacontenttype:
          type: string
          description: 'media-type that describes the event payload encoding, must be "application/json" for CAMARA APIs'
          example: application/json
        data:
          type: object
          description: Event details payload described in each CAMARA API and referenced by its type
        subject:
          description: Describes the subject of the event in the context of the event producer (identified by source).
          type: string
        time:
          $ref: "#/components/schemas/DateTime"
      discriminator:
        propertyName: 'subject'
        mapping:
            EventRoamingStatus: '#/components/schemas/EventRoamingStatus'
            EventRoamingChangeCountry : '#/components/schemas/EventRoamingChangeCountry'
            EventRoamingOff: '#/components/schemas/EventRoamingOff'
            EventRoamingOn: '#/components/schemas/EventRoamingOn'
            EventSubscriptionEnds : '#/components/schemas/EventSubscriptionEnds'`
```

If we use the 'type'  it will be something like this for the extension which seems to us not a good swagger design:

``` yaml
      discriminator:
        propertyName: 'subject'
        mapping:
            org.camara.api.device-status.RoamingStatusEvent: '#/components/schemas/EventRoamingStatus'
            org.camara.api.device-status.RoamingChangeCountryEvent : '#/components/schemas/RoamingChangeCountry'
            org.camara.api.device-status.RoamingOffEvent: '#/components/schemas/EventRoamingOff'
            org.camara.api.device-status.RoamingOnEvent: '#/components/schemas/EventRoamingOn'
            org.camara.api.device-status.SubscriptionEndsEvent : '#/components/schemas/EventSubscriptionEnds
```

WDYT?

Reply:
Understood the point of using "subject" as a discriminator in order to model for both the subscriptions and the events. 
In the second approach propertyName would be "type".

As "type" has reverse-DNS format, it is "long" name and it is a matter of Look&Feel.

And the constraint of doing that is that "subject" needs to be settled as mandatory in our model, because a field being used as discriminator  has to be mandatory (now i understand that this model allows in the case of the subscription to indicate the related information based on eventType, I thought this only refers to event_type and additional info needed for subscription).

So basically, we could do this using "type" without the need to use this additional field.


Regarding "type" format, the comment about using UPPER_SNAKE_CASE was in the line of doing something like that:

org.camara.api.device-status.ROAMING_STATUS (so the eventType is clearly identified in the context)

        mapping:
            org.camara.api.device-status.ROAMING_STATUS: '#/components/schemas/EventRoamingStatus'

Reply:
`type` is clearly intended to be used in CloudEvents as the identifier of the type of event everywhere, and it determines the content of data, so it is the right choice for discriminator. I think this would not be discussed if type didn't require to be expressed in reverse-DNS format, which is not the best name for a schema. But discriminator's mapping comes to solve this, so we can choose whatever schema name is right, without needing to distort the CloudEvents spec.

`subject` is optional precisely because is not needed in every situation. It is not even required to have the same value for all events of the same type. We may discuss if it is needed in CAMARA, but it should not be used as discriminator.

Reply:
Thanks @jlurien for the effort in checking CloudEvents documentation.
Let's move in the direction you suggest.

I will change accordingly the guideline by using `type` as discriminator and not `subject`.

About `subject` probably we could keep this one for later and not define any usage as of now in the guideline.
cc: @rartych